### PR TITLE
 Fix preservation/filtering of exceptions 

### PR DESCRIPTION
**Pull request by aronisstav**
This fixes a regression in OTP 20, caused by the new form of exceptions returned from processes spawned with proc_lib.

While at it, add travis-ci test builds, including Dialyzer, and fix some Dialyzer warnings.

### CODE REVIEW

**Comment by aronisstav** [comment]
Travis-ci needs to be enabled for this repository by an admin.

**Review by aronisstav — COMMENTED** [review]
The results of a Travis test run are available here: https://travis-ci.org/aronisstav/circuit_breaker/builds/398247457

File: src/circuit_breaker.erl [hunk 1]
```diff
@@ -252,6 +255,14 @@ do_call(Service, CallFun, CallTimeout, ResetFun, ResetTimeout, Thresholds) ->
                     ResetTimeout, Thresholds)
   end.
 
+preserve_exception(CallFun) ->
+  try
+    CallFun()
+  catch
+    Class:Reason ->
+      exit({raise, Class, Reason, erlang:get_stacktrace()})
```
Comment (zmstone) [inline src/circuit_breaker.erl:275]: it's deprecated in otp 21. rather avoid using it at all ?

**Comment by zmstone** [comment]
added this project to travis.
the it should trigger a new build when you update this PR.

File: src/circuit_breaker.erl (hunk 1 above)
Comment (aronisstav) [reply src/circuit_breaker.erl:275]: I'd rather keep the possibility of preserving the stacktrace.

I could write two versions of this, using the new `OTP_RELEASE` macro that is introduced in OTP 21:
```
-ifdef(OTP_RELEASE).
%% version with Class:Reason:Stacktrace
-else.
%% current version
-endif.
```
Would that be good enough?

Comment (zmstone) [reply src/circuit_breaker.erl:275]: yes, that'll work.

Comment (aronisstav) [reply src/circuit_breaker.erl:275]: Done

**Review by zmstone — APPROVED** [review]

**Review by k32 — APPROVED** [review]

**Comment by aronisstav** [comment]
@zmstone anything else to be done here before this is merged?

**Comment by zmstone** [comment]
1.0.1

**Pull request merged**